### PR TITLE
feat(fgs): the resource supports reserved instance configuration

### PR DESCRIPTION
**Pull request by wuzhuanhong**
**What this PR does / why we need it**:

The resource supports reserved instance configuration.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. support function reserved instance configuration.
2. support related document and acceptance test.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/fgs TESTARGS='-run TestAccFgsV2Function_reservedInstance'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/fgs -v -run TestAccFgsV2Function_reservedInstance -timeout 360m -parallel 4
=== RUN   TestAccFgsV2Function_reservedInstance_version
=== PAUSE TestAccFgsV2Function_reservedInstance_version
=== RUN   TestAccFgsV2Function_reservedInstance_alias
=== PAUSE TestAccFgsV2Function_reservedInstance_alias
=== CONT  TestAccFgsV2Function_reservedInstance_version
=== CONT  TestAccFgsV2Function_reservedInstance_alias
--- PASS: TestAccFgsV2Function_reservedInstance_alias (24.98s)
--- PASS: TestAccFgsV2Function_reservedInstance_version (39.50s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/fgs       39.554s
```

### CODE REVIEW

File: docs/resources/fgs_function.md [hunk 1]
```diff
@@ -315,6 +315,17 @@ The following arguments are supported:
 
 * `log_stream_name` - (Optional, String) Specifies the name of the LTS stream.
 
+* `qualifier_name` - (Optional, String) Specifies the version name or alias name.
+
+-> If the parameter is alias name, The format is as follows **!xx**.
```
Comment (deer-hang) [inline docs/resources/fgs_function.md]:
```suggestion
  -> If the parameter is alias name, the format is **!xx**.
```
Two spaces are required in front, please check the other parts.

File: docs/resources/fgs_function.md [hunk 2]
```diff
@@ -315,6 +315,17 @@ The following arguments are supported:
...
+
+* `reserved_instance_count` - (Optional, Int) Specifies the number of reserved instance.
+
+* `idle_mode` - (Optional, Bool) Specifies whether to enable the idle mode.
+
+* `tactics_config` - (Optional, List) Specifies the auto scaling policy for reserved instance.
+  The [object](#functiongraph_tactics_config) structure is documented below.
```
Comment (deer-hang) [inline docs/resources/fgs_function.md]:
```suggestion
  The [tactics_config](#functiongraph_tactics_config) structure is documented below.
```
check other parts.

File: docs/resources/fgs_function.md [hunk 3]
```diff
@@ -352,6 +363,45 @@ The `aliases` block supports:
 
 * `description` - (Optional, String) Specifies the description of the version alias.
 
+<a name="functiongraph_tactics_config"></a>
+The `tactics_config` block supports:
+
+* `cron_configs` - (Optional, List) Specifies the lists of scheduled policy configurations.
```
Comment (deer-hang) [inline docs/resources/fgs_function.md]:
```suggestion
* `cron_configs` - (Optional, List) Specifies the list of scheduled policy configurations.
```
Please check the other parts.

File: docs/resources/fgs_function.md [hunk 4]
```diff
@@ -381,7 +435,8 @@ $ terraform import huaweicloud_fgs_function.test <id>
 
 Note that the imported state may not be identical to your resource definition, due to the attribute missing from the
 API response. The missing attributes are:
-`app`, `func_code`, `agency`, `tags"`.
+`app`, `func_code`, `agency`, `tags"`, `package`, `idle_mode`, `qualifier_name`,  `qualifier_type`,
+`reserved_instance_count`, `tactics_config`.
```
Comment (deer-hang) [inline docs/resources/fgs_function.md]: Field `qualifier_type` is Computed, it is unnecessary to ignore it.
Please check the ignore fields. The field qualifier_name has a return value, why should ignore be added?

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go [hunk 5]
```diff
@@ -731,6 +893,75 @@ func updateFunctionVersions(client *golangsdk.ServiceClient, d *schema.ResourceD
...
+		Count:         utils.Int(d.Get("reserved_instance_count").(int)),
+		IdleMode:      utils.Bool(d.Get("idle_mode").(bool)),
+		TacticsConfig: buildTracticsConfigs(d.Get("tactics_config").([]interface{})),
+	}
+
+	_, err := function.UpdateReservedInstanceConfig(client, opts)
+	if err != nil {
+		return fmt.Errorf("error updating function reversed instance: %s", err)
+	}
+	return err
```
Comment (deer-hang) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]:
```suggestion
	return nil
```

File: huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go [hunk 6]
```diff
@@ -962,3 +962,184 @@ resource "huaweicloud_fgs_function" "test" {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
+		ProviderFactories: acceptance.TestAccProviderFactories,
+		CheckDestroy:      rc.CheckResourceDestroy(),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccFgsV2Function_reservedInstance_alias(name),
+				Check: resource.ComposeTestCheckFunc(
+					rc.CheckResourceExists(),
+					resource.TestCheckResourceAttr(resourceName, "qualifier_name", "!alias_demo"),
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go]: Please check whether `qualifier_name` can be changed

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go [hunk 7]
```diff
@@ -656,6 +799,25 @@ func resourceFgsFunctionRead(_ context.Context, d *schema.ResourceData, meta int
...
 		d.Set("versions", versionConfig),
 	)
+
+	reservedInstances, err := getReservedInstanceConfig(fgsClient, d)
+	if err != nil {
+		mErr = multierror.Append(mErr, err)
+	} else if len(reservedInstances) > 0 {
+		reservedInstance := reservedInstances[0]
+		qualifierName := reservedInstance.QualifierName
+		if (reservedInstance.QualifierType) == "alias" && reservedInstance.QualifierName != "" {
```
Comment (deer-hang) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: Why not configure an Optional type field of qualifier_type？

File: huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go [hunk 8]
```diff
@@ -962,3 +962,184 @@ resource "huaweicloud_fgs_function" "test" {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
+		ProviderFactories: acceptance.TestAccProviderFactories,
+		CheckDestroy:      rc.CheckResourceDestroy(),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccFgsV2Function_reservedInstance_step1(name),
+				Check: resource.ComposeTestCheckFunc(
+					rc.CheckResourceExists(),
+					resource.TestCheckResourceAttr(resourceName, "qualifier_name", "latest"),
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go]: Please check whether qualifier_name can be changed

File: docs/resources/fgs_function.md [hunk 9]
```diff
@@ -352,6 +363,45 @@ The `aliases` block supports:
...
+
+* `name` - (Required, String) Specifies the name of scheduled policy configuration.
+
+* `cron` - (Required, String) Specifies the cron expression.
+
+* `count` - (Required, Int) Specifies the number of reserved instance to which the policy belongs.
+
+-> The number of reserved instances must be greater than or equal to the number of reserved instances in the basic configuration.
+
+* `start_time` - (Required, Int) Specifies the effective timestamp of policy.
```
Comment (deer-hang) [inline docs/resources/fgs_function.md]: Please specify the unit of the timestamp and give an example.
Same to `expired_time`

File: docs/resources/fgs_function.md [hunk 10]
```diff
@@ -352,6 +363,45 @@ The `aliases` block supports:
...
+* `cron_configs` - (Optional, List) Specifies the lists of scheduled policy configurations.
+  The [object](#functiongraph_cron_configs) structure is documented below.
+
+* `metric_configs` - (Optional, List) Specifies the lists of metric policy configurations.
+  The [object](#functiongraph_metric_configs) structure is documented below.
+
+<a name="functiongraph_cron_configs"></a>
+The `cron_configs` block supports:
+
+* `name` - (Required, String) Specifies the name of scheduled policy configuration.
```
Comment (deer-hang) [inline docs/resources/fgs_function.md:397]: Add supported character types and length limits for name.

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go [hunk 11]
```diff
@@ -320,6 +322,24 @@ func ResourceFgsFunctionV2() *schema.Resource {
 				Description: "The versions management of the function.",
 			},
 			"tags": common.TagsSchema(),
+			"qualifier_name": {
```
Comment (deer-hang) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: The newly added Optional type fields all have return values. Can Computed be added?

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go [hunk 12]
```diff
@@ -606,6 +695,60 @@ func parseFunctionVersions(client *golangsdk.ServiceClient, functionUrn string)
...
+	reservedInstances, err := function.ListReservedInstanceConfigs(c, opts)
+	if err != nil {
+		return nil, fmt.Errorf("error getting list of the function reserved instance config: %s", err)
+	}
+
+	urn := getReservedInstanceUrn(d, d.Get("urn").(string))
+	result := make([]function.ReservedInstancePolicy, 0, len(reservedInstances))
+	for _, v := range reservedInstances {
+		if v.FunctionUrn == urn {
+			result = append(result, v)
```
Comment (deer-hang) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: Does this need to be added with break?

File: docs/resources/fgs_function.md (hunk 1 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: fixed

File: docs/resources/fgs_function.md (hunk 2 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: fixed

File: docs/resources/fgs_function.md (hunk 3 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: fixed

File: docs/resources/fgs_function.md (hunk 4 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: **qualifier_name ** is changed Computed, the qualifier_name  obtained during the import is empty, the value cannot be get in API request.

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go (hunk 5 above)
Comment (wuzhuanhong) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: fixed

File: huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go (hunk 6 above)
Comment (wuzhuanhong) [reply huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go]: fixed

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go (hunk 7 above)
Comment (wuzhuanhong) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: fixed, add qualifier_type parameter

File: huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go (hunk 8 above)
Comment (wuzhuanhong) [reply huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_function_test.go]: fixed， update case added

File: docs/resources/fgs_function.md (hunk 9 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: fixed

File: docs/resources/fgs_function.md (hunk 10 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md:397]: fixed

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go (hunk 11 above)
Comment (wuzhuanhong) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: If this parameter is not configured when wearing the item resource, they have no value.

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go (hunk 12 above)
Comment (wuzhuanhong) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_function.go]: fixed

File: docs/resources/fgs_function.md [hunk 13]
```diff
@@ -352,6 +355,78 @@ The `aliases` block supports:
...
+  Set this parameter, you can refer to this [document](https://support.huaweicloud.com/intl/en-us/usermanual-functiongraph/functiongraph_01_0908.html).
+
+* `count` - (Required, Int) Specifies the number of reserved instance to which the policy belongs.
+  The valid value ranges from `0` to `1,000`.
+
+  -> The number of reserved instances must be greater than or equal to the number of reserved instances in the basic configuration.
+
+* `start_time` - (Required, Int) Specifies the effective timestamp of policy. The unit is `s`, Example: **1740560074**.
+
+* `expired_time` - (Required, Int) Specifies the expiration timestamp of the policy. The unit is `s`, Example: **1740560074**
```
Comment (Lance52259) [inline docs/resources/fgs_function.md]:
```suggestion
* `start_time` - (Required, Int) Specifies the effective timestamp of policy. The unit is `s`, e.g. **1740560074**.

* `expired_time` - (Required, Int) Specifies the expiration timestamp of the policy. The unit is `s`, e.g. **1740560074**.
```

File: docs/resources/fgs_function.md [hunk 14]
```diff
@@ -352,6 +355,78 @@ The `aliases` block supports:
...
+The `tactics_config` block supports:
+
+* `cron_configs` - (Optional, List) Specifies the list of scheduled policy configurations.
+  The [cron_configs](#functiongraph_cron_configs) structure is documented below.
+
+* `metric_configs` - (Optional, List) Specifies the list of metric policy configurations.
+  The [metric_configs](#functiongraph_metric_configs) structure is documented below.
+
+  -> If you want to use the `metric_configs` parameter, please open a service ticket to enable this function. Refer to
+  [How to submit a service ticket](https://support.huaweicloud.com/intl/en-us/usermanual-ticket/topic_0065264094.html)
```
Comment (Lance52259) [inline docs/resources/fgs_function.md]:
```suggestion
  -> If you want to use the `metric_configs` parameter, please open a service ticket to enable this function. Refer to
     the documentation [how to submit a service ticket](https://support.huaweicloud.com/intl/en-us/usermanual-ticket/topic_0065264094.html).
```

File: docs/resources/fgs_function.md [hunk 15]
```diff
@@ -391,7 +466,7 @@ resource "huaweicloud_fgs_function" "test" {
   ...
   lifecycle {
     ignore_changes = [
-      app, func_code, agency, tags,
+      app, func_code, agency, tags, package
```
Comment (Lance52259) [inline docs/resources/fgs_function.md:469]:
```suggestion
      app, func_code, agency, tags, package,
```

File: docs/resources/fgs_function.md (hunk 14 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: fixed

File: docs/resources/fgs_function.md (hunk 13 above)
Comment (wuzhuanhong) [reply docs/resources/fgs_function.md]: fixed

**Pull request merged**